### PR TITLE
Handle malformed yaml files

### PR DESCRIPTION
**Pull request by jeremylinlin**
- This PR fixes issue https://github.com/operator-framework/operator-courier/issues/16
- Log with error when a malformed YAML is being parsed
- Log with info when valid YAML has been parsed
- Suppress error traceback when calling the CLI

### CODE REVIEW

File: operatorcourier/build.py [hunk 1]
```diff
@@ -3,14 +3,14 @@
 
 class BuildCmd():
     def _get_empty_bundle(self):
-         return dict(
+        return dict(
```
Comment (kevinrizza) [inline operatorcourier/build.py]: There are no actual changes in this file. Please remove this formatting only change as it just adds noise to the commit log.

File: operatorcourier/identify.py [hunk 2]
```diff
@@ -1,16 +1,27 @@
...
-        if "packageName" in operatorArtifact:
-            return "Package"
-        elif "kind" in operatorArtifact:
-            if operatorArtifact["kind"] in ("ClusterServiceVersion", "CustomResourceDefinition"):
-                return operatorArtifact["kind"]
-    raise ValueError('Courier requires valid CSV, CRD, and Package files')
+    try:
+        operatorArtifact = safe_load(operatorArtifactString)
+    except MarkedYAMLError:
+        logger.error("Invalid YAML file")
```
Comment (kevinrizza) [inline operatorcourier/identify.py]: Can we mention which file couldn't be parsed here?

File: operatorcourier/identify.py [hunk 3]
```diff
@@ -1,16 +1,27 @@
...
+        print('Courier requires valid input YAML files')
```
Comment (kevinrizza) [inline operatorcourier/identify.py]: Please avoid using print() and instead only use the logger methods.

File: operatorcourier/identify.py [hunk 4]
```diff
@@ -1,16 +1,27 @@
...
+        exit(0)
```
Comment (kevinrizza) [inline operatorcourier/identify.py]: Rather than `exit(0)` here, can we raise an exception like later in the file?

File: operatorcourier/validate.py [hunk 5]
```diff
@@ -33,16 +33,11 @@ def _bundle_validation(self, bundle):
...
         validationDict[self.csvKey] = self._type_validation(bundleData, self.csvKey, self._csv_validation, True)
         validationDict[self.pkgsKey] = self._type_validation(bundleData, self.pkgsKey, self._pkgs_validation, True)
 
-        valid = True
-        for key, value in validationDict.items():
-            if value is False:
-                valid = False
-        
-        return valid
+        return False not in validationDict.values()
```
Comment (kevinrizza) [inline operatorcourier/validate.py]: How is this change related to the malformed yaml? Please limit this pr to a change related to the comments mentioned in the pr and commit message.

File: tests/test_api.py [hunk 6]
```diff
@@ -14,11 +14,14 @@ def test_make_bundle(directory, expected):
...
-(["tests/test_files/bundles/api/bundle1/crd.yml",
-"tests/test_files/bundles/api/bundle1/csv.yaml",
-"tests/test_files/bundles/api/bundle1/packages.yaml"],
-"tests/test_files/bundles/api/bundle1/results/bundle.yaml"),
+    ([
+         "tests/test_files/bundles/api/bundle1/crd.yml",
+         "tests/test_files/bundles/api/bundle1/csv.yaml",
+         "tests/test_files/bundles/api/bundle1/packages.yaml"
+     ],
+     "tests/test_files/bundles/api/bundle1/results/bundle.yaml"),
```
Comment (kevinrizza) [inline tests/test_api.py]: Same comment as above about limiting prs to changes related to the commit message and purpose.

File: tests/test_files/invalid_malformed_parser_error.yaml [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+          required:
+          - targetNamespace
+          - packages
+          properties:
+            targetNamespace:
+              type: string
+              description: The namespace where the operators will be enabled
+            packages:
+              type: string
+              description: Comma separated list of operator(s) without spaces which will be enabled in the target namespace
```
Comment (kevinrizza) [inline tests/test_files/invalid_malformed_parser_error.yaml]: NIT: Can we use a consistent naming style for filenames in this folder?

File: tests/test_files/invalid_malformed_scanner_error.yaml [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+          required:
+          - targetNamespace
+          - packages
+          properties:
+            targetNamespace:
+              type: string
+              description: The namespace where the operators will be enabled
+            packages:
+              type: string
+              description: Comma separated list of operator(s) without spaces which will be enabled in the target namespace
```
Comment (kevinrizza) [inline tests/test_files/invalid_malformed_scanner_error.yaml]: NIT: Can we use a consistent naming style for filenames in this folder?

File: operatorcourier/identify.py (hunk 4 above)
Comment (scoheb) [reply operatorcourier/identify.py]: or at least exit non-zero...in addition, with #27 ... this should be a critical error in the json output. @awgreene

Comment (jeremylinlin) [reply operatorcourier/identify.py]: @scoheb I have discussed with @kevinrizza. Instead of raising an exception, we will just log error messages to the output. And this will be made consistent in the entire program.

Comment (MartinBasti) [reply operatorcourier/identify.py]: Please never ever call `exit()` in any python module outside CLI processing. It will cause immediate exit from any app using `operator-courier` as its part

File: operatorcourier/identify.py (hunk 2 above)
Comment (jeremylinlin) [reply operatorcourier/identify.py]: This may require changes in multiple function signatures. I instead log all YAMLs that are parsed successfully.

File: operatorcourier/identify.py (hunk 4 above)
Comment (jeremylinlin) [reply operatorcourier/identify.py]: I now only exit(1) in the main entry of the CLI, if any error bubbles up to this level.

File: operatorcourier/cli.py [hunk 9]
```diff
@@ -2,11 +2,17 @@
...
 def main():
     """Generate the CLI bits
     """
-    parser = _CliParser()
-    parser.parse()
+    try:
+        parser = _CliParser()
+        parser.parse()
+    except Exception:  # Exception raised to CLI should not be thrown to users,
+        exit(1)        # it should just be captured by logs
```
Comment (kevinrizza) [inline operatorcourier/cli.py]: I still don't understand why we need to use the `exit()` syscall here. Can't we just print a message and return after printing?

File: operatorcourier/cli.py [hunk 10]
```diff
@@ -32,17 +38,16 @@ def parse(self):
         args = parser.parse_args(sys.argv[1:2])
         if not hasattr(self, args.command):
             parser.error('Unrecognized command')
```
Comment (kevinrizza) [inline operatorcourier/cli.py]: why remove whitespace here? this seems like noise.

File: operatorcourier/cli.py (hunk 9 above)
Comment (jeremylinlin) [reply operatorcourier/cli.py]: What I'm thinking is that
  - we should not make exit call within modules, which we have discussed previously
  - the reason why I didn't just print a message and return is that, by doing this the exit status will be 0. But I want the user to know that the CLI program did not exit correctly.

Any thoughts on that?

File: operatorcourier/cli.py (hunk 10 above)
Comment (jeremylinlin) [reply operatorcourier/cli.py]: I'll remove those then. I think my IDE did this, but I'll make sure this doesn't happen in the future.

File: operatorcourier/cli.py (hunk 9 above)
Comment (kevinrizza) [reply operatorcourier/cli.py]: Right, this is the cli. This looks fine.

Comment (MartinBasti) [inline operatorcourier/cli.py]: I'm not avare that argparser is giving traceback for user errors, only for developer errors.

This case just hides unexpected errors, you at least needs
```
except Exception as e:
   sys.exit(str(e))
```

File: operatorcourier/cli.py [hunk 11]
```diff
@@ -2,11 +2,17 @@
 import sys
 from operatorcourier import api
 
+sys.tracebacklimit = 0  # this suppresses stacktrace of the CLI tool in the command line
```
Comment (MartinBasti) [inline operatorcourier/cli.py]: GL with bug reports without stack traces

File: operatorcourier/identify.py [hunk 12]
```diff
@@ -1,16 +1,34 @@
...
-            return "Package"
-        elif "kind" in operatorArtifact:
-            if operatorArtifact["kind"] in ("ClusterServiceVersion", "CustomResourceDefinition"):
-                return operatorArtifact["kind"]
-    raise ValueError('Courier requires valid CSV, CRD, and Package files')
+    try:
+        operatorArtifact = safe_load(operatorArtifactString)
+    except MarkedYAMLError:
+        logger.error("Courier requires valid input YAML files")
+        raise ValueError('Courier requires valid input YAML files')
```
Comment (MartinBasti) [inline operatorcourier/identify.py]: Pro tip:
```python
msg = "Courier requires valid input YAML files"
logger.error(msg)
raise ValueError(msg)
```

File: operatorcourier/identify.py [hunk 13]
```diff
@@ -1,16 +1,34 @@
...
+        if operatorArtifact is not None:
+            if "packageName" in operatorArtifact:
+                artifact_type = "Package"
+                artifact_name = operatorArtifact['packageName']
+            elif "kind" in operatorArtifact:
+                if operatorArtifact["kind"] in ("ClusterServiceVersion", "CustomResourceDefinition"):
+                    artifact_type = operatorArtifact["kind"]
+                    artifact_name = operatorArtifact['metadata']['name']
+            if artifact_type is not None:
+                logger.info(f'Parsed {artifact_type}: {artifact_name}')
```
Comment (MartinBasti) [inline operatorcourier/identify.py]: No format strings (`f'...'`)! It is supported only in py 3.7+. We need older python versions too at least 3.6.

Also this logging is not lazy, please use `logger.info('Parsed %s: %s', artifact_type, artifact_name)` instead.

File: operatorcourier/identify.py [hunk 14]
```diff
@@ -1,16 +1,34 @@
...
 
 def get_operator_artifact_type(operatorArtifactString):
-    """get_operator_artifact_type takes a yaml string and determines if it is one of the 
+    """get_operator_artifact_type takes a yaml string and determines if it is one of the
     expected bundle types: ClusterServiceVersion, CustomResourceDefinition, or Package.
 
     :param operatorArtifactString: Yaml string to type check
     """
-    operatorArtifact = yaml.safe_load(operatorArtifactString)
-    if isinstance(operatorArtifact, dict):
```
Comment (MartinBasti) [inline operatorcourier/identify.py]: You don't need this check anymore? What if yaml contains only list?

File: operatorcourier/identify.py (hunk 13 above)
Comment (MartinBasti) [reply operatorcourier/identify.py]: Ah sorry format strings are supported in py 3.6, but there is still that lazy not lazy logging.

**Comment by jeremylinlin** [comment]
@MartinBasti Thanks for your comments. I have updated the PR.

File: tests/test_files/invalid.malformed.parser.error.yaml [hunk 15]
```diff
@@ -0,0 +1,54 @@
+- apiVersion: apiextensions.k8s.io/v1beta1
```
Comment (awgreene) [inline tests/test_files/invalid.malformed.parser.error.yaml]: Could you add a comment highlighting where the yaml is malformed? This will help debugging issues in the future.

File: tests/test_files/invalid.malformed.scanner.error.yaml [hunk 16]
```diff
@@ -0,0 +1,54 @@
+apiVersion: apiextensions.k8s.io/v1beta1
```
Comment (awgreene) [inline tests/test_files/invalid.malformed.scanner.error.yaml]: Could you add a comment highlighting where the yaml is malformed? This will help debugging issues in the future.
